writer.go

```diff
@@ -595,13 +596,24 @@ func (w *writer) run() {
 				resch = append(resch, wm.res)
 				mustFlush = len(batch) >= w.batchSize
 			}
+			if !forceBatchRunning {
+				forceBatch.Reset(w.batchTimeout)
+				forceBatchRunning = true
+			}
 
-		case now := <-ticker.C:
-			mustFlush = now.Sub(lastFlushAt) > w.batchTimeout
+		case <-forceBatch.C:
+			mustFlush = true
+			forceBatchRunning = false
 		}
 
 		if mustFlush {
-			lastFlushAt = time.Now()
+			if forceBatchRunning {
+				stopped := forceBatch.Stop()
+				forceBatchRunning = false
+				if !stopped {
```

Comment:
I would prefer to see this line moved up (and possibly even collapsed into `if stopped := forceBatch.Stop(); !stopped`) so that we put all the logic for timer management together and then toggle the `forceBatchRunning` boolean.  I think it's easier to reason about if we don't mix the two.

Reply:
Field moved up & all uses of `forceBatch` have been changed to `batchTimer`.